### ISSUE TITLE
maximize_directivity: stop_gradient(P_rad) yields wrong-sign gradients for power-changing DoFs (valid only for scale/shape-preserving DoFs)

### ISSUE THREAD

**Issue by bk-squared** [bk-squared/rfx#129]
> Reported from a separate analysis workspace; **no rfx file has been modified.**
> This is a diagnosis plus a suggested back-compat patch for maintainers to
> consider. The repro/validation scripts referenced below are the reporter's and
> can be shared on request; the quotient-rule decomposition, the measured
> numbers, and the patch are inlined so the report is self-verifiable without them.

## Summary

`maximize_directivity` (`rfx/optimize_objectives.py`, lines ~284–371) computes
the directivity ratio as

```python
directivity = u_target / (jax.lax.stop_gradient(p_rad) + 1e-30)
```

The `stop_gradient` on the `P_rad` denominator drops the `−U·P′/P²` term of the
quotient rule. The docstring justifies this as a "scale-invariant" trick (issue
#32): for a *shape-only* DoF, `P_rad` is approximately constant so the dropped
term is negligible.

That assumption is **violated by any DoF that changes total radiated power** —
exactly the class of inverse-design knobs used for antennas: a conductor /
topology cell that adds or removes metal lowers **both** `U_target` and
`P_rad`. For that class the dropped term is the *same magnitude class* as the
kept term and **opposes it**, so the AD gradient comes out with the **wrong
sign** relative to finite differences. A wrong-signed gradient makes the
optimizer climb away from the directivity maximum with no warning.

Measured on a power-changing in-place PEC strip DoF (CPU, tiny grid):

| objective | AD gradient | central FD | sign | rel_err |
|-----------|-------------|------------|------|---------|
| rfx default `−U/stop_gradient(P)` | **+9.483e-3** | −1.860e-3 | **FLIPPED** | 1.000 |
| log-ratio fix `−(log U − log P)` | −6.299e-3 | −6.294e-3 | correct | 0.000 |

The raw far-field adjoint itself is **exact** — this is purely the objective's
gradient formula, not the NTFF/PEC physics (see "Scope" below).

## Affected APIs / DoF classes

`maximize_directivity` (and its alias `maximize_directivity_ratio`) used as the
objective in `optimize()` / `topology_optimize()` whenever the design DoF
**changes total radiated power**:

- **Conductors / PEC** — `topology_optimize(... material_fg="pec")`, or
  `pec_occupancy`-based parametric optimization (Yagi director/reflector
  offsets and lengths, patch trimming, parasitic elements). Adding/removing
  metal scatters more/less power, so `P_rad` moves with the DoF. **Gradient
  always corrupted; sign can flip** — verified on an in-place PEC-occupancy DoF;
  parametric offset/length DoFs are *not* guaranteed to flip and must be
  screened per-DoF (in our gate a moving-offset DoF stayed same-sign but
  magnitude-wrong).
- **Lossy / absorbing materials** — any DoF that changes `sigma` and thus
  dissipated vs radiated power. **Affected.**
- **Dielectric (eps) reshaping** — `P_rad` moves less, so the *sign* often
  survives, but the dropped term still corrupts the *magnitude*: in our control
  the eps DoF gave AD=−3.240e-1 vs FD=−3.466e-2 (rel_err 0.807) under the rfx
  default. **Magnitude wrong; convergence/step-scaling degraded.**
- **Pure shape-preserving DoFs** (the original issue-#32 target, `P_rad ≈`
  const) — unaffected; this is the regime the `stop_gradient` was designed and
  validated for (#32).

Net effect: gradient-based inverse design with a directivity objective over
power-changing DoFs can converge in the wrong direction with no error. The
dropped `−U·P′/P²` term is always the same magnitude class as the kept `U′/P`
term (so the gradient magnitude is always wrong), and it flips the *sign* only
when it exceeds and opposes the kept term (see decomposition below). Whether a
specific DoF flips or merely loses accuracy is operating-point dependent and
should be screened per-DoF with an AD-vs-central-FD check.

## Root cause — quotient-rule decomposition (proof)

`D = U/P`. The exact derivative w.r.t. a design parameter `θ` is

```
dD/dθ = (U′·P − U·P′) / P²  =  U′/P  −  U·P′/P²
```

`stop_gradient(P)` keeps only the first term, `U′/P`, and **drops** `−U·P′/P²`.
The loss is `−D`, so:

```
d(−D)/dθ  true            = −(U′P − U·P′)/P²
d(−D)/dθ  with stop_grad  = −(U′/P)
dropped term              = + U·P′/P²
```

When `P′` and `U′` have the same sign (both fall as metal is removed) and
`U·P′/P²` exceeds `U′/P` in magnitude, the kept and true derivatives have
**opposite signs**.

### Measured decomposition (this is the proof)

From `scripts/tap_diag_ratio_stopgrad.py` (`results/.../diag_ratio_stopgrad.json`),
at the base point on the in-place PEC DoF:

```
U0 = 3.6586e-27      P0 = 1.2364e-26
dU = −1.1726e-28     dP = −4.7413e-28        (both U and P FALL with the DoF)

d(−U/P)_true = −(dU·P0 − U0·dP)/P0²          = −1.8637e-03
d(−U/P)_sg   = −(dU/P0)                        = +9.4841e-03   ← wrong sign
dropped term =  U0·dP/P0²                       = −1.1348e-02
```

`d(−U/P)_true (−1.86e-3) == central FD (−1.860e-3)` to 3 digits → the **full
quotient is correct**; the stop_gradient version (+9.48e-3) is the wrong sign
because the dropped term (−1.13e-2) is larger than and opposes the kept term
(+9.48e-3 contribution). This is a clean, fully-explained sign flip, not noise.

### Why not just delete `stop_gradient`?

Because `U` and `P` are ~1e-27..1e-26 in rfx's spectral NTFF convention, the
naive full quotient `−U/(P+1e-30)` divides near the 1e-30 floor and its
gradient evaluates to **NaN** (verified: candidate L2 in
`scripts/tap_diag_logratio_fix.py` gives `AD=nan`). The denominator floor that
makes the forward value finite does not make the backward pass finite. The fix
has to be both *full-quotient* and *scale-robust*.

## Proposed fix — fully-differentiated log-ratio objective

Optimize the **log** of the ratio:

```
loss = −(log(U_target + eps) − log(P_rad + eps))
grad = −(U′/(U+eps) − P′/(P+eps))
```

This is the log-derivative `d(log U − log P) = U′/U − P′/P`, which equals the
true `dD/dθ` up to the positive factor `1/D` (hence sign-correct and monotone in
the directivity — same optimum). It is scale-invariant (an overall NTFF magnitude scale cancels exactly, preserving the
issue-#32 property that motivated the original change), and is NaN-safe because
each log argument is independently floored (so each ratio is `O(1)`, never a
1e-27/1e-30 blow-up). The objective is monotone in the directivity, so the
optimum is unchanged.

To **preserve existing behavior**, add it as a new flag (and/or a new
discoverable factory `maximize_directivity_logratio`) rather than changing the
default. Unified-diff-style snippet against `rfx/optimize_objectives.py`:

```diff
@@ def maximize_directivity(
     theta_target: float,
     phi_target: float,
     *,
     n_theta: int = 37,
     n_phi: int = 73,
+    log_ratio: bool = False,
+    eps: float = 1e-30,
 ) -> Callable:
@@
-    ``stop_gradient`` is applied to the P_rad denominator: since the
-    ratio is scale-invariant, any shape-preserving scaling leaves the
-    directivity unchanged; letting the denominator carry gradient would
-    only introduce noise and NaN risk when P_rad is near zero early in
-    optimization.
+    ``stop_gradient`` is applied to the P_rad denominator ONLY in the
+    default (``log_ratio=False``) mode. WARNING: that mode drops the
+    -U*P'/P^2 quotient term and therefore yields WRONG-SIGN gradients for
+    any DoF that changes total radiated power (conductors / PEC topology /
+    lossy materials). For those DoFs pass ``log_ratio=True``, which
+    optimizes -(log U - log P): the full, scale-invariant quotient
+    grad = U'/U - P'/P, NaN-safe via independent floors, no stop_gradient.
@@
         p_rad_phi = jnp.trapezoid(integrand, theta_hemi, axis=1)
         p_rad = jnp.trapezoid(p_rad_phi, phi_hemi, axis=1)  # (n_freqs,)
 
-        directivity = u_target / (jax.lax.stop_gradient(p_rad) + 1e-30)
-        # Minimizing -D = maximizing directivity; average across freqs
-        return -jnp.mean(directivity)
+        if log_ratio:
+            # Full, NaN-safe quotient: grad = U'/U - P'/P. Floors are applied
+            # independently so each log argument is O(1) (the spectral NTFF
+            # scale is ~1e-27; a single shared 1e-30 floor on U/P makes the
+            # backward pass NaN). No stop_gradient -> correct sign for
+            # power-changing DoFs.
+            log_dir = (jnp.log(jnp.maximum(u_target, eps))
+                       - jnp.log(jnp.maximum(p_rad, eps)))
+            return -jnp.mean(log_dir)
+        # Legacy scale-invariant ratio (shape-only DoFs ONLY — see warning).
+        directivity = u_target / (jax.lax.stop_gradient(p_rad) + eps)
+        return -jnp.mean(directivity)
@@
 maximize_directivity_ratio = maximize_directivity
+
+
+def maximize_directivity_logratio(
+    theta_target: float,
+    phi_target: float,
+    *,
+    n_theta: int = 37,
+    n_phi: int = 73,
+    eps: float = 1e-30,
+) -> Callable:
+    """Directivity objective with the full (sign-correct) quotient gradient.
+
+    Equivalent to ``maximize_directivity(..., log_ratio=True)``. Prefer this
+    for power-changing DoFs (PEC/topology/lossy); see the sign-flip warning on
+    ``maximize_directivity``.
+    """
+    return maximize_directivity(
+        theta_target, phi_target, n_theta=n_theta, n_phi=n_phi,
+        log_ratio=True, eps=eps,
+    )
```

## Validation of the fix

`scripts/tap_diag_logratio_fix.py` implements the three candidate objectives
inline and screens each with the canonical gate harness
`experiments/_gate_harness.py :: ad_fd_sign_screen` (the same AD-vs-central-FD
AD-vs-central-FD sign gate used throughout this investigation). Results
(`results/tap_moving_boundary_gate/diag_logratio_fix.json`):

| DoF | objective | AD | FD | sign_ok | rel_err |
|-----|-----------|----|----|---------|---------|
| in-place PEC (power-changing) | L1 rfx `−U/stop_grad(P)` | +9.483e-3 | −1.860e-3 | **False** | 1.000 |
| in-place PEC | L2 naive `−U/(P+eps)` | **NaN** | −1.860e-3 | False | NaN |
| in-place PEC | **L3 log-ratio (fix)** | −6.299e-3 | −6.294e-3 | **True** | 0.000 |
| dielectric eps (control) | L1 rfx | −3.240e-1 | −3.466e-2 | True | 0.807 |
| dielectric eps | **L3 log-ratio (fix)** | −8.544e-2 | −8.545e-2 | **True** | 0.000 |

Verdict logged by the script: `rfx-default-sign-flips=True`,
`logratio-sign-correct(pec&eps)=True`. The log-ratio objective is sign-correct
**and** magnitude-exact on the power-changing PEC DoF and on the eps control;
the rfx default is sign-flipped on PEC and magnitude-wrong on eps; the naive
no-stop_gradient quotient is NaN. Radiation-pattern artifact (broadside-peaked
phi=0 cut at the screen base point) rendered to
`results/tap_moving_boundary_gate/logratio_pattern.png`.

## Reproduction

All CPU, tiny grid (`F0=3 GHz`, `dx=λ/10`, 4λ domain, 200 steps), a few minutes
each. Run from `papers/rfx-tap/`:

```bash
python scripts/tap_diag_ratio_stopgrad.py     # quotient-rule decomposition + R1/R2 (the proof numbers above)
python scripts/tap_diag_ntff_pec_adjoint.py   # localizes flip to the ratio, NOT the PEC->NTFF adjoint
python scripts/tap_diag_logratio_fix.py        # screens the log-ratio fix with ad_fd_sign_screen + pattern PNG
```

- `scripts/tap_diag_ratio_stopgrad.py` — decomposes `U′`, `P′` separately and
  compares `d(−U/P)_true` (== central FD) vs `d(−U/P)_sg` (rfx). Source of the
  `U0/P0/dU/dP` numbers above.
- `scripts/tap_diag_ntff_pec_adjoint.py` — D1 raw `|E_θ|²` broadside adjoint
  (exact) vs D2 directivity ratio (flips) on the *same* PEC DoF; D3 eps control.
  Establishes the fault is the objective, not the far-field/occupancy adjoint.
- `scripts/tap_diag_logratio_fix.py` — implements L1/L2/L3 and runs the
  `ad_fd_sign_screen` gate; emits the verdict JSON and the pattern PNG.
- Shared primitives (a maintainer must copy these alongside the diag scripts to
  reproduce): `scripts/tap_gate_moving_boundary.py` (grid, NTFF box, in-place
  PEC/eps occupancy builders, forward passes) and
  `experiments/_gate_harness.py` (`ad_fd_sign_screen`). Self-contained
  Python + rfx + JAX; no other rfx-tap-private deps.

## Scope — what is NOT broken

The raw far-field adjoint is exact and is **not** implicated:
`tap_diag_ntff_pec_adjoint.py` D1 (raw broadside `|E_θ|²` w.r.t. the same
in-place PEC DoF, no ratio, no stop_gradient) agrees with central FD in sign and
magnitude. `apply_pec_occupancy` and the NTFF surface DFT differentiate
correctly. The defect is entirely in the directivity objective's gradient
formula (`stop_gradient(P_rad)`), so the fix is local to
`optimize_objectives.py`.

## Environment

- rfx version: **1.6.3** (`rfx.__version__`)
- rfx-ref HEAD: **`b62cbaa`** (`v1.6.3-324-gb62cbaa`), branch
  `feat/waveguide-tjunction-meep-ref`
- JAX, float32, **CPU** (`JAX_PLATFORM_NAME=cpu`), tiny grid — no GPU needed
- Source under test: `rfx/optimize_objectives.py` `maximize_directivity`
  (lines ~284–371), `directivity = u_target / (jax.lax.stop_gradient(p_rad) + 1e-30)`
- Validation gate: `papers/rfx-tap/experiments/_gate_harness.py :: ad_fd_sign_screen`
```

## Honest caveats

- Validated on a single power-changing in-place PEC DoF and one eps control DoF
  at one base point on a tiny CPU grid. The mechanism (quotient-rule term drop)
  is base-point-independent, but the *full-Yagi multi-DoF* and *real multi-cell
  topology* sign-correctness claim is a prediction to be re-screened with
  `ad_fd_sign_screen` on the actual §5 design vectors before it is relied on in
  the paper.
- The `eps` magnitude rel_err under the rfx default (0.807) shows the dropped
  term degrades dielectric DoFs too; this is asserted, not separately decomposed
  here.
- Numbers are float32 on CPU; the `true == FD` agreement is to ~3 significant
  figures, which is the expected float32/FD-step precision, not a tighter claim.
